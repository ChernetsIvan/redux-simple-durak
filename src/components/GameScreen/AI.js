import React from "react";
import PropTypes from "prop-types";
import { uniqueId } from "lodash-es";

import InvisibleCard from "./Card/InvisibleCard";
import HiddenCard from "./Card/HiddenCard";

const AI = ({ cards }) => {
  let output = null;
  if (cards.length > 0) {
    let inputCards = [];
    inputCards = cards.concat();
    let countOfCardsThatNotShown = 0; // сколько карт не видно
    let outCountOfCardsThatNotShown = null;
    // Выводим только первые 9 рубашкой кверху карт
    while (inputCards.length > 9) {
      inputCards.pop();
      countOfCardsThatNotShown += 1;
      outCountOfCardsThatNotShown = <b>+{countOfCardsThatNotShown}</b>;
    }

    const outputCards = inputCards.map(() => {
      const randKey = uniqueId();
      return <HiddenCard key={randKey} bootStrapColClass="col-1 ml-3" />;
    });

    output = (
      <div className="container mt-1">
        <div className="row justify-content-center">
          {outputCards}
          <div className="col-1 align-self-center">
            {outCountOfCardsThatNotShown}
          </div>
        </div>
      </div>
    );
  } else {
    output = <InvisibleCard />;
  }
  return (
    <div className="container">
      <div className="row">
        <div className="col-1" />
        <div className="col-10">{output}</div>
        <div className="col-1" />
      </div>
    </div>
  );
};

export default AI;

AI.propTypes = {
  cards: PropTypes.arrayOf(PropTypes.object).isRequired
};
