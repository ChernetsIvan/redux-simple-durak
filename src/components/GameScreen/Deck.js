import React from "react";
import PropTypes from "prop-types";

import Card from "./Card/Card";
import HiddenCard from "./Card/HiddenCard";

const Deck = ({cards}) => {
    var cardsWithoutLast = [];
    cardsWithoutLast = cards.concat();
    let trumpCard = cardsWithoutLast.pop();

    //1)Козырная карта.
    let output_TrumpCard = null;
    if (trumpCard !== undefined) {
      //"Переопределяем" bootstrap-класс col-6 вместо col-1:
      output_TrumpCard = (
        <Card
          rank={trumpCard.rank}
          suit={trumpCard.suit}
          bootStrapColClass="col-6"
          hidden={false}
        />
      );
    }

    //2)Имитация колоды.
    let output_MockDeck = null;
    let output_CountOfCardsInDeck = null;
    if (cardsWithoutLast.length > 0) {
      output_MockDeck = (
        <HiddenCard bootStrapColClass="col-6" />
      );
      output_CountOfCardsInDeck = cardsWithoutLast.length;
    }

    let output = (
      <div className="container">
        <div className="row">
          {output_TrumpCard}
          {output_MockDeck}
        </div>
        <div className="row">
          <div className="col-6" />
          <div className="col-6">
            <div className="container">
              <div className="row justify-content-center">
                <div className="col-auto">
                  <b>{output_CountOfCardsInDeck}</b>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
    return output;
};

export default Deck;

Deck.propTypes = {
  cards: PropTypes.array
};
