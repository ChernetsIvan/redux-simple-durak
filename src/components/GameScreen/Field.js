import React from "react";
import PropTypes from "prop-types";

import { uniqueId } from "lodash-es";

import InvisibleCard from "./Card/InvisibleCard";
import Card from "./Card/Card";

const Field = ({ cards }) => {
  let output = null;

  // Чтобы "не скакала" разметка:
  if (cards.length === 0) {
    output = <InvisibleCard />;
  } else {
    const outCards = cards.map(el => {
      const key = `${el.suit.suit}${el.rank.text}`;
      return (
        <Card
          key={key}
          rank={el.rank}
          suit={el.suit}
          bootStrapColClass={"col-1"}
        />
      );
    });
    output = (
      <div className="container">
        <div className="row">{outCards}</div>
      </div>
    );
  }
  return output;
};

export default Field;

Field.propTypes = {
  cards: PropTypes.arrayOf(PropTypes.object).isRequired
};
