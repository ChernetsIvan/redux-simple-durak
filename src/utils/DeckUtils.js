import { pull, find } from 'lodash-es';

import {cardRanks, cardSuits} from './../constants/GameConstants';

import CardModel from './../models/CardModel';

class DeckUtils{
    //отсортировать входной массив карт по возрастанию isAsc=true
    //(убыванию-> isAsc=false)  по параметру 'power'
    static sortInputDeckByPower(inputDeck, isAsc){
        inputDeck.sort(function(a,b){
            if(isAsc){
                return a.power - b.power;
            }else{
                return b.power - a.power;
            }        
        });
    }

    //Возвращает наполненную всеми 36-ю картами колоду
    static fillCards(trumpSuit){
        let fullDeck = [];
        //Устанавливает power для каждой карты по принципу:
        //козырь->10*rank.cardValue, НЕкозырь->1*rank.cardValue
        let z = 0;
        let power, coef;               
        cardSuits.forEach(function(element,index, array){
            if(element.suit === trumpSuit.suit){
                coef = 10;
            }else{
                coef = 1;
            }
            cardRanks.forEach(function(element2, index2, array2){
                power = coef * element2.cardValue;
                fullDeck.push(
                    new CardModel(z, element2, element, power)
                );
                z++;
            });
        });
        return fullDeck;
    }

    //Возвращает перемешанную колоду
    static shuffle(inputDeck){
        var newFullDeck = [];
        while(inputDeck.length > 0){
            //Случайный индекс в inputDeck
            let randIndex = Math.random() * (inputDeck.length-1);
            randIndex = Math.round(randIndex);

            let card = inputDeck[randIndex];
            newFullDeck.push(card);

            inputDeck.splice(randIndex, 1);
        }
        return newFullDeck;
    }

    //Перемещает первую попавшуюся карту с козырной мастью
    //в конец колоды 
    static moveAnyCardWithTrumpSuitToTailOfFullDeck(trumpSuit, fullDeck){
        let card = find(fullDeck, (card) => {
            return card.suit.suit === trumpSuit.suit;
        });
        pull(fullDeck, card);
        fullDeck.push(card);
    }

    //взять из НАЧАЛА (!) колоды fullDeck и положить в inputDeck столько карт, 
    //чтобы в inputDeck в итоге было 6 карт
    static giveUpToSixCards(fullDeck, inputDeck){
        //если хотя бы одно из условий не выполняется, в цикл не заходим:
        while(inputDeck.length < 6 && fullDeck.length > 0){
            let card = fullDeck.shift();
            inputDeck.push(card);
        }
    }

    //Возвращает козырную масть (не карту!) для будущей игры
    static chooseSuitForTrumpCardInFuture(){
        let randIndex = Math.random() * (cardSuits.length-1);
        randIndex = Math.round(randIndex);
        return cardSuits[randIndex];
    }
}

export default DeckUtils;