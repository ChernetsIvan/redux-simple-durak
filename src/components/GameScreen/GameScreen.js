import React from "react";
import PropTypes from "prop-types";

// Компонент-представление
import Field from "./Field";

// Компоненты-контейнеры
import StatusContainer from "./../../containers/GameScreen/StatusContainer";
import AIContainer from "./../../containers/GameScreen/AIContainer";
import DeckContainer from "./../../containers/GameScreen/DeckContainer";
import PlayerContainer from "./../../containers/GameScreen/Player/PlayerContainer";

const GameScreen = ({ onBeginGameClick, aiField, playerField }) => (
  <div>
    <div className="container  my-1">
      <div className="row justify-content-between">
        <div className="col-auto">
          <button className="btn btn-secondary mr-3" onClick={onBeginGameClick}>
            Новая игра
          </button>
        </div>
        <StatusContainer />
        <div className="col-2" />
      </div>
    </div>
    <div>
      <AIContainer />
      <div className="container">
        <div className="row align-items-center">
          <div className="col-10">
            <Field cards={aiField} />
            <Field cards={playerField} />
          </div>
          <div className="col-2">
            <DeckContainer />
          </div>
        </div>
      </div>
      <PlayerContainer />
    </div>
  </div>
);

export default GameScreen;

GameScreen.propTypes = {
  onBeginGameClick: PropTypes.func.isRequired,
  aiField: PropTypes.arrayOf(PropTypes.object).isRequired,
  playerField: PropTypes.arrayOf(PropTypes.object).isRequired
};
