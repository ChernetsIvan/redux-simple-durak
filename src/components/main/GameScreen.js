import React, { Component } from 'react';
import PropTypes from 'prop-types';

//Классы-компоненты
import './../Card/Card.css' // Подключаем единожды, чтобы не объявлять в каждом из components!
import AI from './../AI';
import Player from './../Player';
import Deck from './../Deck';
import Field from './../Field';
import Status from './../Status';

class GameScreen extends Component {    
    render() {
        return (
            <div>
                <div className="container  my-1">
                    <div className="row justify-content-between">
                        <div className="col-auto">
                            <button
                                className="btn btn-secondary mr-3" 
                                onClick={this.props.handleBeginGameClick}>
                                
                                Новая игра
                            </button>
                        </div>
                        <Status 
                            gameMode={this.props.gameMode}
                            computerCards={this.props.computerCards}
                            playerCards={this.props.playerCards}
                            fullDeck={this.props.fullDeck}
                            firstStart={this.props.firstStart} />
                        <div className="col-2"></div>
                    </div>  
                </div>
                <div>
                    <div className="container">
                        <div className="row">
                            <div className="col-1"></div>
                            <div className="col-10">
                                <AI cards={this.props.computerCards} />
                            </div>
                            <div className="col-1"></div>
                        </div>
                    </div>
                    <div className="container">
                        <div className="row align-items-center">
                            <div className="col-10">
                                <Field cards={this.props.aiField} />                
                                <Field cards={this.props.playerField} />
                            </div>
                            <div className="col-2">
                                <Deck cards={this.props.fullDeck} />
                            </div>                                                       
                        </div>
                    </div>
                    <div className="container">
                        <div className="row">
                            <div className="col-1"></div>
                            <div className="col-10">
                                <Player 
                                    cards={this.props.playerCards}                            
                                    onPrevClick={this.props.onPrevClick}
                                    onNextClick={this.props.onNextClick}
                                    playerStartInd={this.props.playerStartInd}
                                    playerEndInd={this.props.playerEndInd}
                                    handlePlayerMove={this.props.handlePlayerMove} />
                            </div>
                            <div className="col-1"></div>
                        </div>
                        <div className="row justify-content-center">
                            <div className="col-auto">
                                
                            </div> 
                        </div>
                    </div>                                 
                </div>
            </div>
        );
    }
}

export default GameScreen;

GameScreen.propTypes = {
    handleBeginGameClick: PropTypes.func.isRequired
};