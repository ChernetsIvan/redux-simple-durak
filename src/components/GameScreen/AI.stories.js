import React from "react";

import { storiesOf } from "@storybook/react";

import { AI } from "./AI";
import DeckUtils from "./../../utils/DeckUtils";

const cards = DeckUtils.fillCards({ suit: "Ð‘" });

storiesOf("AI", module)
  .add("with 1 card", () => <AI cards={cards.slice(0, 1)} />)
  .add("with 7 cards", () => <AI cards={cards.slice(5, 12)} />)
  .add("with 20 cards", () => <AI cards={cards.slice(16, 36)} />);
