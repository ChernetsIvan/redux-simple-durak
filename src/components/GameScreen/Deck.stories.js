import React from "react";

import { storiesOf } from "@storybook/react";

import { Deck } from "./Deck";
import DeckUtils from "./../../utils/DeckUtils";

const cards = DeckUtils.fillCards({ suit: "Ð§" });

storiesOf("Deck", module)
  .addDecorator(story => (
    <div className="container">
      <div className="row">
        <div className="col-2">{story()}</div>
      </div>
    </div>
  ))
  .add("with 10 cards", () => <Deck cards={cards.slice(0, 10)} />)
  .add("with 30 cards", () => <Deck cards={cards.slice(5, 35)} />)
  .add("with 1 card", () => <Deck cards={cards.slice(35, 36)} />);
