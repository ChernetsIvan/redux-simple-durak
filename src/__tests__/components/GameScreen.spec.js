import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-15";
import toJson from "enzyme-to-json";

import GameScreen from "./../../components/GameScreen/GameScreen";
import DeckUtils from "./../../utils/DeckUtils";

Enzyme.configure({ adapter: new Adapter() });

describe("GameScreen shallow render test", () => {
  test("test", () => {
    const emptyFunc = () => {};
    const cards = DeckUtils.fillCards({ suit: "Ð‘" });

    const component = shallow(
      <GameScreen
        onBeginGameClick={emptyFunc}
        aiField={cards.slice(0, 5)}
        playerField={cards.slice(0, 5)}
      />
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
