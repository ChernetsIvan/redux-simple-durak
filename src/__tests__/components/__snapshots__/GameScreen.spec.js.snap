// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameScreen shallow render test test 1`] = `
<div>
  <div
    className="container  my-1"
  >
    <div
      className="row justify-content-between"
    >
      <div
        className="col-auto"
      >
        <button
          className="btn btn-secondary mr-3"
          onClick={[Function]}
        >
          Новая игра
        </button>
      </div>
      <Connect(Status) />
      <div
        className="col-2"
      />
    </div>
  </div>
  <div>
    <Connect(AI) />
    <div
      className="container"
    >
      <div
        className="row align-items-center"
      >
        <div
          className="col-10"
        >
          <Field
            cards={
              Array [
                CardModel {
                  "id": 0,
                  "power": 1,
                  "rank": Object {
                    "cardValue": 1,
                    "text": "6",
                  },
                  "suit": Object {
                    "suit": "Ч",
                  },
                },
                CardModel {
                  "id": 1,
                  "power": 2,
                  "rank": Object {
                    "cardValue": 2,
                    "text": "7",
                  },
                  "suit": Object {
                    "suit": "Ч",
                  },
                },
                CardModel {
                  "id": 2,
                  "power": 3,
                  "rank": Object {
                    "cardValue": 3,
                    "text": "8",
                  },
                  "suit": Object {
                    "suit": "Ч",
                  },
                },
                CardModel {
                  "id": 3,
                  "power": 4,
                  "rank": Object {
                    "cardValue": 4,
                    "text": "9",
                  },
                  "suit": Object {
                    "suit": "Ч",
                  },
                },
                CardModel {
                  "id": 4,
                  "power": 5,
                  "rank": Object {
                    "cardValue": 5,
                    "text": "10",
                  },
                  "suit": Object {
                    "suit": "Ч",
                  },
                },
              ]
            }
          />
          <Field
            cards={
              Array [
                CardModel {
                  "id": 0,
                  "power": 1,
                  "rank": Object {
                    "cardValue": 1,
                    "text": "6",
                  },
                  "suit": Object {
                    "suit": "Ч",
                  },
                },
                CardModel {
                  "id": 1,
                  "power": 2,
                  "rank": Object {
                    "cardValue": 2,
                    "text": "7",
                  },
                  "suit": Object {
                    "suit": "Ч",
                  },
                },
                CardModel {
                  "id": 2,
                  "power": 3,
                  "rank": Object {
                    "cardValue": 3,
                    "text": "8",
                  },
                  "suit": Object {
                    "suit": "Ч",
                  },
                },
                CardModel {
                  "id": 3,
                  "power": 4,
                  "rank": Object {
                    "cardValue": 4,
                    "text": "9",
                  },
                  "suit": Object {
                    "suit": "Ч",
                  },
                },
                CardModel {
                  "id": 4,
                  "power": 5,
                  "rank": Object {
                    "cardValue": 5,
                    "text": "10",
                  },
                  "suit": Object {
                    "suit": "Ч",
                  },
                },
              ]
            }
          />
        </div>
        <div
          className="col-2"
        >
          <Connect(Deck) />
        </div>
      </div>
    </div>
    <Connect(Player) />
  </div>
</div>
`;
